### PR TITLE
Implement wav2gloss (ACL 2024)

### PR DESCRIPTION
**Pull request by juice500ml**
## What?

- Implement ASR1 recipe for wav2gloss task
    - Implement `text_prev` for ASR1 (for injecting task tokens in multi-task training)
    - Tested single-task and multi-task models for the wav2gloss task
    - Tested two different underlying SSL models (XLS-R, WavLM)

## Why?

- To provide a reproducible codebase for ACL 2024 [Wav2Gloss: Generating Interlinear Glossed Text from Speech](https://arxiv.org/abs/2403.13169)

## See also

- Other baseline for the ACL 2024 paper: https://github.com/juice500ml/finetune_owsm

### CODE REVIEW

File: egs2/TEMPLATE/asr1/db.sh [hunk 1]
```diff
@@ -190,6 +190,7 @@ KIRITAN=
 NAMINE=
 KATHBATH=downloads
 GRAMVAANI=downloads
+WAV2GLOSS=downloads
```
Comment (ftshijt) [inline egs2/TEMPLATE/asr1/db.sh:193]: Please add in the egs2/README.md

File: egs2/TEMPLATE/asr1/asr.sh
Comment (sw005320) [inline egs2/TEMPLATE/asr1/asr.sh:1]: Is this change necessary? 
I don't want to make this larger.

Comment (sw005320) [reply egs2/TEMPLATE/asr1/asr.sh:1]: This change affects many part.
If you want to keep this, that's fine. 
But you need to make a test to cover this case.

File: egs2/TEMPLATE/asr1/db.sh (hunk 1 above)
Comment (juice500ml) [reply egs2/TEMPLATE/asr1/db.sh:220]: Sorry, missed this, will fill up now!

Comment (juice500ml) [reply egs2/TEMPLATE/asr1/db.sh:220]: Fixed in 16d7f201772d4c50ab64311353f7f4bf7bff6ed1

File: egs2/TEMPLATE/asr1/asr.sh
Comment (juice500ml) [reply egs2/TEMPLATE/asr1/asr.sh:1]: I tried to make the existing recipes not broken via having default parameters. Nevertheless, I agree with your concern completely, so I'll add the test for this.

**Comment by sw005320** [comment]
@ftshijt, can you review this again?
There are some changes in the ASR template. 
So, let's check it carefully.

**Comment by juice500ml** [comment]
I'm also somewhat lost on the test failures... Now, after some refactoring, there is little to no edits in the beam search implementation. However, the test I added is now failing. It'll be either (i) tests are wrong OR (ii) existing implementation on hyp_primer was wrong.

**Comment by sw005320** [comment]
I have never encountered this kind of issues
https://github.com/espnet/espnet/actions/runs/9989159778/job/27607159902?pr=5449#step:8:171

I'll rerun the CI.

**Comment by juice500ml** [comment]
image
Fails on my local also, only on transformer lm

**Comment by sw005320** [comment]
@ juice500ml, this PR includes too many additions.
Can you first make a PR only with the recipe part?

**Comment by juice500ml** [comment]
@sw005320 , I think that's a better idea! I did removed some of the beam search refactoring codes on this branch (moved to https://github.com/espnet/espnet/pull/5846). However, I couldn't remove the edits on `asr.sh` and `asr_inference.py`, because, without them, the recipe won't work. To cover the edits made in `asr_inference.py`, I added a little bit more test code just to make sure.

**Comment by juice500ml** [comment]
Weird that test configuration is failing 😢 Any idea why?

**Comment by ftshijt** [comment]
> Weird that test configuration is failing 😢 Any idea why?

from the log
`Error: egs2/TEMPLATE/asr1/cmd.sh and egs2/wav2gloss/asr1/cmd.sh differ: To solve: for f in egs2/*/*/cmd.sh; do cp egs2/TEMPLATE/asr1/cmd.sh ${f}; done`

**Comment by juice500ml** [comment]
Ohhh, I see! Just modified it. Hope the tests go through this time :)

**Comment by sw005320** [comment]
@ftshijt, do you have other ideas for throwing the previous text?
For me, this PR has too many changes for the main template to support one recipe.

**Comment by ftshijt** [comment]
> @ftshijt, do you have other ideas for throwing the previous text? For me, this PR has too many changes for the main template to support one recipe.

One option is that we can have a copy of specific-inference code/customized stages stored in local?

**Comment by juice500ml** [comment]
Closed per discussion last week.
